### PR TITLE
Dpctl c api constructors added, dpctl/include restructured 

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
Refactored `dpctl/include` layout. Instead of containing all files from `libsyclinterface/include` it introduces substructure:

```
   dpctl/
      include/
         syclinterface/
              *.h   # all the files that used to be in dpctl/include
         syclinterface.h   
         dpctl_capi.h
         dpctl4pybind11.hpp
```

The newly added files are residing in new folder `dpctl/apis/include` from which `build_backend.py` copies then into `dpctl/include`

Modified `_backend.pxd` to include `"syclinterface/"` in the path. 

Modified `build_backend.py` to reassemble `dpctl/include` from `install/include` and `dpctl/apis/include`.

Added `make_Sycl*` C-API functions to create `SyclDevice`, `SyclQueue`, `SyclContext` and `SyclEvent` from corresponding opaque reference. Used `get_device_ref` and `make_SyclDevice` to construct pybind11 type caster directly mapping `dpctl.SyclDevice` Python object to `sycl::device` in pybind11.

Modified examples files to use new headers, and use casters.

Verified that all Cython extensions and Pybind11 extensions are building and are running as expected.

### CODE REVIEW

File: dpctl/_sycl_context.pyx [hunk 1]
```diff
@@ -485,3 +485,12 @@ cdef api DPCTLSyclContextRef get_context_ref(SyclContext ctx):
     :class:`dpctl.SyclContext` instance.
     """
     return ctx.get_context_ref()
+
+
+cdef api SyclContext make_SyclContext(DPCTLSyclContextRef CRef):
```
Comment (diptorupd) [inline dpctl/_sycl_context.pyx:490]: Why are these changes here?

File: dpctl/apis/include/dpctl4pybind11.hpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+#include <CL/sycl.hpp>
+#include <pybind11/pybind11.h>
+
+namespace py = pybind11;
+
+namespace pybind11
+{
+namespace detail
+{
+template <> struct type_caster<sycl::queue>
```
Comment (diptorupd) [inline dpctl/apis/include/dpctl4pybind11.hpp:38]: Please add docstrings. So that we can start generating these files in API docs under maybe programmer's guide.

File: dpctl/_sycl_context.pyx (hunk 1 above)
Comment (oleksandr-pavlyk) [reply dpctl/_sycl_context.pyx:490]: Because without them pybind11 casters can not be written. One needs 2 way street to go from Python object to c++ class, and back.

File: dpctl/apis/include/dpctl_capi.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+/*
+ * Function to import dpctl and make C-API functions available.
+ * C functions can use dpctl's C-API functions without linking to
+ * shared objects defining this symbols, if they call `import_dpctl()`
+ * prior to using those symbols.
+ */
+void import_dpctl(void)
+{
+    import_dpctl___sycl_device();
```
Comment (diptorupd) [inline dpctl/apis/include/dpctl_capi.h:49]: Oh My! that is some function name :smile:

File: dpctl/apis/include/syclinterface.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file includes all the headers of syclinterface/
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+// clang-format off
```
Comment (diptorupd) [inline dpctl/apis/include/syclinterface.h:27]: why format off?

File: dpctl/apis/include/syclinterface.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#include "syclinterface/dpctl_vector.h"
+#include "syclinterface/dpctl_utils.h"
+#include "syclinterface/dpctl_sycl_device_selector_interface.h"
+#include "syclinterface/dpctl_sycl_context_interface.h"
+#include "syclinterface/dpctl_sycl_device_interface.h"
+#include "syclinterface/dpctl_sycl_event_interface.h"
+#include "syclinterface/dpctl_sycl_platform_interface.h"
+#include "syclinterface/dpctl_sycl_queue_interface.h"
+#include "syclinterface/dpctl_sycl_usm_interface.h"
+
```
Comment (diptorupd) [inline dpctl/apis/include/syclinterface.h:40]: spurious new line?

File: examples/cython/sycl_buffer/use_sycl_buffer.cpp [hunk 6]
```diff
@@ -28,7 +28,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "use_sycl_buffer.h"
-#include "dpctl_sycl_types.h"
+#include "syclinterface.h"
```
Comment (diptorupd) [inline examples/cython/sycl_buffer/use_sycl_buffer.cpp:31]: should the header be called `dpctl_sycl_interface.h`?

File: dpctl/apis/include/syclinterface.h (hunk 4 above)
Comment (oleksandr-pavlyk) [reply dpctl/apis/include/syclinterface.h:27]: Because clang would reorder includes and include C-API headers before including `dpctl_sycl_types` and then the opaque reference identifiers would be unknown and compilation fails.

File: examples/cython/sycl_buffer/use_sycl_buffer.cpp (hunk 6 above)
Comment (oleksandr-pavlyk) [reply examples/cython/sycl_buffer/use_sycl_buffer.cpp:31]: Ok, will make the change.

File: dpctl/apis/include/syclinterface.h (hunk 5 above)
Comment (oleksandr-pavlyk) [reply dpctl/apis/include/syclinterface.h:40]: It was deliberate, to separate `_interface.h` from `_manager.h`, but I am going to remove it.

File: examples/cython/sycl_buffer/use_sycl_buffer.cpp (hunk 6 above)
Comment (oleksandr-pavlyk) [reply examples/cython/sycl_buffer/use_sycl_buffer.cpp]: done

File: dpctl/apis/include/syclinterface.h (hunk 5 above)
Comment (oleksandr-pavlyk) [reply dpctl/apis/include/syclinterface.h]: Pushed the change to remove the line and rename the file.

File: dpctl/apis/include/dpctl4pybind11.hpp (hunk 2 above)
Comment (oleksandr-pavlyk) [reply dpctl/apis/include/dpctl4pybind11.hpp:43]: Done. Added comments to document what's going on.

**Comment by oleksandr-pavlyk** [comment]
@densmirn @Alexander-Makaryev Most likely `dpnp` is going to be affected by this change. Once this PR is merged, the downstream (relative to `dpctl`) libraries will need to ``#include "dpctl_sycl_interface.h"`` instead of including individual headers, such as `dpctl_sycl_types.h` and `dpctl_sycl_queue_manager.h`.

@napetrov Please forward this to relevant people working on `scikit-learn-intelex`.

**Comment by oleksandr-pavlyk** [comment]
@diptorupd This change is going to cause some disruption in downstream code:

1. https://github.com/IntelPython/numba-dppy/blob/main/numba_dppy/dpctl_iface/usm_shared_allocator_ext.c#L30
2. https://github.com/IntelPython/dpnp/blob/master/dpnp/backend/src/queue_sycl.hpp#L44
3. https://github.com/intel/scikit-learn-intelex/blob/master/src/dpctl_interop/daal_context_service.cpp#L20

I am not sure what `DPPLQueueMgr_GetCurrentQueue` is in scikit-learn-intelex, and I did not see it defined anywhere in the project. Perhaps @napetrov could clarify.

But perhaps to minimize the disruption, we could temporarily add symbolic links in `dpctl/include/` for some commonly used header files?

**Comment by diptorupd** [comment]
> But perhaps to minimize the disruption, we could temporarily add symbolic links in `dpctl/include/` for some commonly used header files?

No please. It is a minimal change that we have enough time to fix in downstream. I will suggest that we keep the change out of 0.12 and then fix it.

**Comment by oleksandr-pavlyk** [comment]
> I will suggest that we keep the change out of 0.12 and then fix it.


I am afraid I need parts of this PR (support for dpctl4pybind11) in 0.12.

How about we add a header file (`dpctl/include/dpctl_sycl_queue_manager.h` that contains `#error` with the text explaining what to do.

This header is to be removed before 0.12 tag is created.

**Review by Alexander-Makaryev — DISMISSED** [review]

**Comment by oleksandr-pavlyk** [comment]
> I would therefore merge this to master once this change is added to downstream projects.

Actually, testing locally, there are no conflicts between this branch and the work in #683
